docs/src/design/proposals/Registration.md

```diff
@@ -0,0 +1,518 @@
+  zero or more instances of objects of type `X` which have been registered,
+  using a `compositor` function provided as a constructor argument (or via
+  a subsequent `compose` call.)
+  * `register(factory : (function () : X), [options] : RegistrationOptions)`
+    registers an object (as returned by the provided `factory` function)
+    with this provider. Evaluation of the provided `factory` will be
+    deferred until the first `get()` call to the `Provider`.
+  * `compose(compositor : (function (X[]) : S), [options] : RegistrationOptions)`
+    introduces a new strategy for converting an array of registered objects
+    of type `X` into a single instance of an object of type `S`. The
+    highest-priority `compositor` that has been registered in this fashion
+    will be used to assemble the provided object (before decoration)
+* `ServiceProvider<S>` extends `CompositeProvider<S, S>` and provides analogous support
+  for the _composite services_
+  pattern used throughout Open MCT (which, in turn, is a superset of the
+  functionality needed for plain services.)
+* `Registry<T>` extends `CompositeProvider<T, T[]>` and provides analogous support for
+  _extension categories_, also
+  used ubiquitously through Open MCT.
+* `RegistrationOptions` provides a location for additional
+  registration-related metadata to be provided.
+  * `priority` is used to control the ordering and/or preference of
+    factories, compositors, and/or decorators,
+    [as in the existing Framework Layer.](http://nasa.github.io/openmctweb/guide/#priority)
+
+# Examples
+
+The following examples are provided to illustrate the intended usage of
+the Registration API. Particular attention is given to obeying and
+utilizing the "dependency injection as code style"
+[decision from the API Redesign](APIRedesign.md#decisions).
+
+## Building Applications on Open MCT
+
+Applications built using Open MCT are expected to extend the `MCT` base
+class and should typically self-install distinguishing plugins during
+the constructor call. Any pre-installed plugins that are undesirable
+should also be uninstalled at this point (to support such usage,
+`MCT` should expose instances of any installed plugins that are
+considered optional.)
+
+For example:
+
+```
+define(['mct', './plugins'], function (mct, plugins) {
+  var MCT = mct.MCT;
+
+  function Variant() {
+    MCT.call(this);
+
+    this.install(new plugins.SomePlugin());
+    this.install(new plugins.SomeOtherPlugin(this.core));
+
+    this.uninstall(this.plugins.persistence.localStorage);
+  }
+
+  Variant.prototype = Object.create(MCT.prototype);
+
+  return Variant;
+});
+```
+
+Running an application build using Open MCT then typically looks like:
+
+```
+define(['./Variant], function (Variant) {
+  new Variant().run();
+});
+```
+
+## Writing Plugins
+
+A plugin for Open MCT should inherit from the `Plugin` base class.
+
+Plugins will typically use extension points exposed by other plugins;
```

Comment:
What do we gain by having two plugin architectures?  
Well, what have we gained by having one plugin architecture?  

Nothing.

In fact, we have lost things:  
* We have lost hours of development time due to closely coupled code and poor platform architecture that was hidden behind the facade of "everything is a plugin".
* We have lost the equivalent of a full year of development time due to the steep learning curve and impossible to follow architecture that resulted in 3 developers being unable to jump in and contribute code.[1]
* We have yet to attract developers to the project on it's merits alone.  I have personally tried to sell developers on the tool only to show them the docs and plugin system and watch them develop cold feet and back track.
* We have lost hours of time just debating the concept of a plugin system in the first place.  

And for all these losses, we have not realized any benefits from this plugin system that we could not have realized without it.  We went to water the garden and decided to purchase a fire truck to replace our hose.  Unfortunately, fire trucks get gallons to the mile and require tons of maintenance and labor to keep running[2].

So yea, ultimately, I don't belive we need an internal plugin architecture.  Javascript is a very flexible language with strong modular concepts (thanks to requirejs which paved the way for browserify and finally the new ES6 module standards), and we should use that to build our application.  We should not add any more concepts to the mix.  We should certainly not invent any new concepts.

I don't believe we need a complex and feature rich external plugin architecture.  We need the simplest interfaces with clean and clear documentation.  We can evolve from there, and should we encounter a real need for a more complicated plugin interface, we should design it when we need it-- which is not right now.

What would I propose?  

Throw away bundles.  Throw away our current generic extension points.  Build a core application which exposes specific APIs.  Define extension points where we know we need them.  No weird descriptive metadata-- just javascript.  

Suddenly, our architecture becomes clear, clean, and easy to explain.  It's simple to describe what we do, what we provide.  Such things are hard to do when our most truthful description of the platform was as a "plugin extension mechanism".  

With clear APIs that do exactly one thing well, It's simple to describe how to extend the platform-- there are no intermediate concepts, the learning process is straightforward and to the point.

Anything in the core platform that would utilize extension points becomes an external plugin: plots, elastic search persistence, etc-- those are all plugins, and they're registered via a dead simple API.  

Things like time conductor-- that is a feature of the platform, and the platform itself accepts some configuration to enable/disable it.

[1] Yes, there were multiple reasons this didn't work out.  But having grown to know these developers, if the platform had helped them achieve anything in a straightforward manner, I believe they would have been eager to work with it.
[2] http://geckodesigns.kinja.com/so-i-bought-a-firetruck-252516685